### ISSUE TITLE
Speed-up fastq reading (by not converting phred scores?)

### ISSUE THREAD

**Issue by jrderuiter** [biocore/scikit-bio#907]
I have been working on some code where I read sequences from a fastq file, perform some operations on these sequences and then write said sequences to a new fastq file. I use the generator interface from io for this, as this allows me to perform the work with low memory overhead for large files. I have noticed however, that a considerable amount of time is being spent on reading/writing fastq sequences, relative to the computation that is being performed on the sequences. This leads to a long runtime on larger (Illumina) fastq files which is undesirable in my case.

From some experiments with my own code I think that a 'considerable' amount of time in reading/writing fastq sequences goes into converting shred scores into a numeric representation (my own code is much faster if I skip this and simply include the quality scores as a string). 

Maybe it would be possible to speedup sequence reading/writing by allowing the caller to skip this kind of conversion if the converted scores are not used anyway. Scores should however still be stored within the sequence object, as writing sequences to a new fastq file should preserve the original scores.

**Comment by anderspitman** [comment]
:+1: I really like this idea. scikit-bio's io system has a lot of potential in conversion pipelines. There may be some issues initially storing the phred scores as text and only verifying/converted if/when they are used. For example, users aren't too surprised if reading a file takes some time, but it might seem odd if reading was really fast and then your first operation was slow. But if we can figure a way to do this that makes sense it would be sweet.

Do you have a rough estimate of the speedup ratio you're seeing?

**Comment by jrderuiter** [comment]
I made a quick notebook to play around with some options here: http://nbviewer.ipython.org/gist/anonymous/9e30c828b08137e678cb. In this notebook I essentially read 1000 fastq sequences from a file and time the reading.

Just using skbio takes around ~55 ms. My 'quick-and-dirty' reader with a simple sequence class reduces this to ~2.5 ms, though it doesn't perform any validation or conversion. Adding conversion increases the time needed to ~20 ms, indicating that conversion does take a substantial amount of time. Interestingly, using the DNASequence class doubles the runtime to 42 ms, of which most time is going into the _set_quality function, indicating that we might be able to win time there as well.

**Comment by jrderuiter** [comment]
Using the skbio's reading code with an adjusted reader that skips conversion and a simpler sequence class with no validation, gives me a speed-up of ~5x. Additional speed-up might be obtained on the reading side, though I don't know how much faster it can get with the required checks and multi-line fastq compatibility. 

On another note, would cythonizing the code provide any benefits? (I have little experience with Cython).

**Comment by gregcaporaso** [comment]
Thanks for looking into this @jrderuiter! 

@jairideout, @ebolyen can you comment? Do you expect your updates to the ``Sequence`` object to achieve this same/similar speed-up? If not, we definitely need to be able to support suppressing validation for applications like this.

**Comment by ebolyen** [comment]
Hey @jrderuiter, one thing that will provide an immediate improvement is if you add the keyword `verify=False` to the read call. Without that, scikit-bio will double check that the file is in fact fasta 
before continuing. 

@gregcaporaso None of our changes to sequence are relevant at this point as it's interface is a numpy array.

In regards to delayed quality parsing, I think that is a good idea, but there is also another option which may make this less of an issue. Currently we are converting with a list and an 'ord' operation. But there is no reason we couldn't vectorize this with numpy which can grab the buffer behind the python string. This would also prevent `_set_quality` from doing any actual work. I am going to play with this option to see if we can get the time somewhere more reasonable. (Thanks for your benchmarks by the way @jrderuiter, this is definitely a huge performance issue!)

**Comment by wasade** [comment]
@ebolyen, I thought the vectorizing was already planned for? I think this is an excellent idea that has potential to offer immediate performance benefits.

@jrderuiter, cython could be great here too but I think we want to make sure the interface is hardened first as it is an increased maintenance burden.

**Comment by ebolyen** [comment]
@jrderuiter Is it possible for you to email me the test file you are using? (I have a gmail account with this same username.) I seem to be unable to reproduce these runtimes on python 2.7 with the latest master. I suspect sequence length has something to do with the discrepancy.

**Comment by ebolyen** [comment]
Actually, looks like we are using different python versions. I will change the kernel to use py3 and try again.

**Comment by jrderuiter** [comment]
With vectorisation you mean using numpy to convert phred scores? We should be able to do that with numpy as follows:

```{python}
def decode_qual(s):
    return np.fromstring(s, dtype='uint8') - 33

def decode_qual2(qual_str, phred_offset=33):
    ac = np.array(qual_str, 'c')
    return ac.view(np.uint8) - phred_offset
```

I haven't been able to time this yet, as timeit is warning me about possible caching of results because of timing differences. Maybe something to do with time needed for memory allocation in bumpy?

**Comment by ebolyen** [comment]
@jrderuiter I have a new branch of skbio that is using essentially that technique (fromstring is much faster, btw). Benchmarking now.

**Comment by jrderuiter** [comment]
OK, nice. I have been using the fromstring variant in my own code as well.

**Comment by ebolyen** [comment]
Really promising results:

Before vectorization:
http://nbviewer.ipython.org/gist/ebolyen/c615d63a1840623bbbd8
After:
http://nbviewer.ipython.org/gist/ebolyen/7e7286e704b1934191d0

Plain skbio:
From 92.5 ms to 48.2 ms

Simple implementation with conversion:
From 50.8 ms to 15.7 ms

I'll issue a PR for this soon. @jrderuiter, we are also in the process of re-factoring our sequence objects to make them faster so that should cut down on some of the cost we are seeing in the benchmarks. Also I agree with @wasade that Cythonizing would be excellent, but we should think carefully about it before we proceed. I think an ideal situation would be some low level Cython parsers that can just pass buffers straight into various sequence objects and have all of the validation/sanity checking handled in Cython/numpy. 

We are still trying to move towards beta however, so our API is slightly more important at this time. But once that is stabilized, performance is the next target!

**Comment by wasade** [comment]
That is awesome!!

**Comment by jrderuiter** [comment]
I guess #908 addresses this issue. The other issues (concerning setting of quality scores on Sequence objects, which is also slow) are being addressed elsewhere right?

**Comment by ebolyen** [comment]
@jrderuiter Not yet, but I think it is worthwhile to add some kind of mechanism to ignore features you don't care about when reading (such as quality).

**Comment by shiffer1** [comment]
Hello All, 
Jai asked me to add a note about how ignoring the quality scores could be useful. I had to download a set of fastq files from a different website than we are used to. This is the ENA the European Nucleotide Archive. The sequences have either been re-encoded in some strange format or corrupted somehow as they contain unicode characters. In any case, I would like to have the option to ignore the phred scores and just reformat the sequences into a fasta. This is due to the fact that the original authors stated the data was already quality filtered. It becomes problematic if you have unicode or other characters in the quality scores as they are specific to a range of phred scores. In fact, almost every program errors when trying to convert to fasta. It would be a nice feature to have a quick fastq to fasta convert ignoring a QUAL output or just filtering characters outside of the range (with some sort of smart offset checking).
Thanks,
Arron

**Comment by wasade** [comment]
@shiffer1, if you haven't reached out, I'm sure EBI would appreciate a note about the corrupt fastqs

**Comment by gregcaporaso** [comment]
I agree on both points: it'd be great to just be able to ignore the quality scores, and contacting ENA about it would be great.

**Comment by shiffer1** [comment]
Hello,
I will try to contact them,but I am not convinced that this isn't a fluke,
or perhaps just operator error on the part of the up-loader.
Arron

**Comment by gregcaporaso** [comment]
Even so, it's good for them to know so they can correct this record.
